File: answers/lintcode/Interleaving String.cpp
//@result 

class Solution {
public:
    /**
     * Determine whether s3 is formed by interleaving of s1 and s2.
     * @param s1, s2, s3: As description.
     * @return: true of false.
     */
    bool isInterleave(string s1, string s2, string s3) {
        // write your code here
		if (s1.length() + s2.length() != s3.length()) {
			return false;
		}
		vector<vector<bool> > dp(s1.length() + 1, vector<bool> (s2.length() + 1, false));
		for (size_t i = 0; i <= s1.length(); ++ i) {
			for (size_t j = 0; j <= s2.length(); ++ j) {
				dp[i][j] = ((0 == i && 0 == j) || (i > 0 && dp[i - 1][j] && s1[i - 1] == s3[i + j - 1]) || (j > 0 && dp[i][j - 1] && s2[j - 1] == s3[i + j - 1]));
			}
		}
		return dp[dp.size() - 1][dp[dp.size() - 1].size() - 1];
    }
};

File: answers/other/coding/Interleaving String.py
class Solution(object):
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        
        if len(s1) + len(s2) != len(s3):
        	return False
        dp = [[False for i in range(len(s2) + 1)] for j in range(len(s1) + 1)]
        for i in range(len(s1) + 1):
        	for j in range(len(s2) + 1):
        		dp[i][j] = (0 == i and 0 == j) or (i >= 1 and s1[i - 1] == s3[i + j - 1] and dp[i - 1][j]) or (j >= 1 and s2[j - 1] == s3[i + j - 1] and dp[i][j - 1])
        #print "debug", dp
        return dp[-1][-1]